### PR TITLE
2 new events (AllowPlayerClaimPermissionEvent, PlayerAbandonClaimEvent) - bis

### PR DESCRIPTION
**Pull request by Vilevar**
Because the IAddonPlugin interface of EtienneDx isn't on the code and I didn't find how I can get the same skills with what there already is, I created 2 new events :  
- AllowPlayerClaimPermissionEvent : It's fired when a method "allowSomething" in the class Claim is called with the parameter "fireEvent" on true (so we also can have without event to avoid loops) and it allows to modify the denial message. The only "special" thing is the permission type (from the enum AllowedPermissionType) which depends on which method from the class Claim called the event (allowEdit or allowBuild or ...) so there is one event for all permission types (a little bit as PlayerInteractEvent for all actions).
- PlayerAbandonClaimEvent : It's fired when a player try to abadon a claim. I created it because ClaimDeletedEvent is too general and called too late (which implies that it cannot be cancelled).

(I'm not sure if it is 1 or 2 features but because it replaces 1 thing (the IAddonPlugin interface) I send it in a single Pull Request.)

### CODE REVIEW

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: Shouldn't be trailing whitespace here, only nitpicking it because of other problems.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: This label usage is pretty bad practice. It makes the code harder to follow and increases indentation for no real benefit - anything done with it could easily be done by adding additional else blocks.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: Can you explain a usage in which one would not want to fire an event? This does not seem like a good design choice.

This applies to all of the trust/permission checks, not going to bother noting the rest because it's pretty clear where they are.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: Same as above, abusing labels when it's easily avoidable.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: More label abuse.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: More label abuse.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: More label abuse.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: Small nitpick again, whitespace.

File: src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java]: Space after if

File: src/main/java/me/ryanhamshire/GriefPrevention/events/AllowPlayerClaimPermissionEvent.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/events/AllowPlayerClaimPermissionEvent.java]: Braces on new lines for this whole class

File: src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java
Comment (Jikoo) [inline src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java]: Braces on new lines for this whole class

**Review by Jikoo — CHANGES_REQUESTED** [review]
I want to preface my review by saying that I appreciate how much work goes into changes like this. Thank you for helping out and submitting what you've done.

I would like to see the `PlayerAbandonClaimEvent` ~~(which should actually likely just be an `AbandonClaimEvent` and support any `CommandSender` so it is also fired when console abandons claims)~~ as a separate PR, there's a lot to unpack here. I'm not going to focus on it here due to the separate nature, but I will say that it looks like you're missing a couple cases, at very least `/abandonallclaims`.

Overall I'm not a fan of the route you've taken. There is currently a LOT of duplicate logic across claim checks that could be reduced, and this only adds more. When I added `ClaimPermission#isGrantedBy(ClaimPermission)`, I planned to revisit claim permission logic and refactor at very least all `ClaimPermission` checks into a single method, if not also PermissionTrust.

/e `ClaimDeletedEvent` handles console deleting claims, that's fine. I definitely recall seeing players with a single claim doing `/abandonallclaims` constantly because they'd place a chest temporarily and then move on.

File: src/main/java/me/ryanhamshire/GriefPrevention/events/AllowPlayerClaimPermissionEvent.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/events/AllowPlayerClaimPermissionEvent.java]: I forgot, sorry

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: Yes, when it calls another allowMethod (such as for parent Claim) it should fire twice for the same test which is useless.
I also think that sometimes we might want to have the result of GP only, without the intervention of other plugins.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: I didn't think about adding "else if" blocks because I wanted to have something as near of the original code as possible but I'm going to fix this (if it's worth it).

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Jikoo) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: That's reasonable. It probably should be private then, addons shouldn't be able to bypass checks.

File: src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java]: I didn't know your plans about revisiting claim permission logic and I'm sorry, I think it's a good idea but now it's not like that (I think...) : I only use the duplicate logic which already were there because I've found that the methods `Claim#allowSomething(Player, [Material])` are more "powerful" that the method `Claim#hasExplicitPermission(Player, ClaimPermission)` because these methods use the other one so that means that setting an event at these methods allows us to do more customizations and therefore I did it. I hope you better understand why I worked that way...

About the `PlayerAbandonClaimEvent` (I'll rename it) I didn't think about /abandonall, I didn't look for each ways to unclaim because I was thinking that it was everytime with the same method so I'll fix it.
But I know that ClaimDeletedEvent is better and handles (I think) everything but as I said, it's fired to late, we can't cancel it so I find it not so useful (and personally I need to cancel some abandons)

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:331]: I've some problems yet : when there is 
```
if(...[1]) {
    if(...[2]) {
        return ... ;
    }
} else { ... }
```
, it is really not the same as 
```
if(...[1] && ...[2]) {
     return ... ;
} else { ... }
```
so I don't know how to do with only else blocks (because there are situations like that, e.g. with the test for admin claims) and it's maybe because I'm a bad developer but I'm still stuck...

Comment (Jikoo) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: I think you're probably just getting too hung up on maintaining a minimal line content diff. When you look at the actual logic, there are few else blocks used, which means that all failed conditions fall through to the next if. That means that if you stack all of the conditions in a single block, you can actually achieve the required behavior with an else.

Examine the following:
```Java
if (condition1) {
  if (condition2) {
    return x;
  }
}
if (condition3) {
  return y;
}
return z;
```
Vs.
```Java
if (condition1 && condition2) {
  return x;
} else if (condition3) {
  return y;
} else {
  return z;
}
```

If it's truly not possible to do in an efficient manner without quick returns, then a helper method should be created instead.

File: src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java
Comment (Jikoo) [reply src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java]: I assume you meant to respond to the main review, so I'll reply there.

**Comment by Jikoo** [comment]
>I didn't know your plans about revisiting claim permission logic and I'm sorry, I think it's a good idea but now it's not like that (I think...) : I only use the duplicate logic which already were there because I've found that the methods Claim#allowSomething(Player, [Material]) are more "powerful" that the method Claim#hasExplicitPermission(Player, ClaimPermission) because these methods use the other one so that means that setting an event at these methods allows us to do more customizations and therefore I did it. I hope you better understand why I worked that way...

I'm not saying you should have used `Claim#hasExplicitPermission`, I'm saying you should consider a new method, something like `Claim#handleAction(Player, ClaimPermission, @Nullable Material)` to remove the heavy duplication of the allow methods.

I think that the various allow methods may not be powerful enough. I would like to see GP wrap the causing event so addons have full control based on any factors imaginable.  For example, imagine that I want to create an addon that allows shearing pink sheep in claims.

Current single-listener GP bypass (sends denial, semi-compatible with other addons):
* Uncancel sheep interaction at lower priority than GP if my conditions are met

Current single-listener GP bypass (no denial, incompatible with other addons):
* Unregister GP's listener for sheep interaction and manually pass it the event if my conditions are not met

Current double-listener GP bypass (no denial, semi-compatible with other addons):
* Ensure listener for sheep is registered at lowest before GP
* Cancel event before GP receives it if my conditions are met so message is not sent
* Uncancel event after GP skips it if my conditions are met

With your implementation (no denial, fully compatible with other addons):
* Ensure listener for sheep is registered at lowest before GP
* Create a flag for the next GP event
* Allow GP's event to occur and remove flag

You can see how while the compatibility with other addons increases, the code complexity required doesn't decrease as much as it could.

Ideal implementation (no denial, fully compatible with other addons):
* Allow GP's event to occur if my conditions are met

**Comment by RoboMWM** [comment]
I haven't read through the diff or comment thread here yet but please keep PRs limited to one PR per feature. Makes both the diff and conversation easier to track and update when given feedback.

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: I think I'll use a helper method because there's an `elseif` with the `condition3` so it is not equivalent (maybe in practice well but I don't want to take the risk).

File: src/main/java/me/ryanhamshire/GriefPrevention/Claim.java
Comment (Vilevar) [reply src/main/java/me/ryanhamshire/GriefPrevention/Claim.java]: I'm not sure to understand (I'm not English-speaking) : with this system the addons can't bypass the checks, they can "only" modify what's the return of the checks. And with the parameter `fireEvent`, the default checks are also done and I don't think that making this method private is a good idea (e.g. if there is 2 addons, they could create an infinite loop) but if you think it's necessary (to do it private) I'll do.

**Comment by Vilevar** [comment]
Ok, I'll create later a new PR for the second event, sorry.

**Comment by Vilevar** [comment]
Ok, I understand the first idea but I'm not sure to understand the example so I'll try something and I'll wait for your feedback.

**Comment by Jikoo** [comment]
This capability has been added in #1006. Addons can listen to and modify the ClaimPermissionCheckEvent to change GP's core behavior.

**Pull request closed without merging**
